### PR TITLE
Screenshot with callbacks

### PR DESCRIPTION
**Pull request by khangtb1**
Maintain the idea of https://github.com/COVESA/wayland-ivi-extension/pull/129
Some improvements:
- Rename the new functions
- Typedef for define function pointer
- Fix the checking filename in callback
- Support non-blocking when filename is empty
- Create functional tests

### CODE REVIEW

File: ivi-layermanagement-api/ilmControl/include/ilm_control.h [hunk 1]
```diff
@@ -365,6 +365,27 @@ ilmErrorTypes ilm_displaySetRenderOrder(t_ilm_display display, t_ilm_layer *pLay
...
+ * \param[in] screen Id of screen where screenshot should be taken
+ * \param[in] filename Optional location where the screenshot should be stored
+ * if the filename is empty and callbacks are available, function will not blocks
+ * \param[in] callback_done Optional callback called when screenshot is acquired
+ * \param[in] callback_error Optional callback called when screenshot acqusition failed
+ * \param[in] user_data Optional callback user data passed in by called
+ * \return ILM_SUCCESS if the method call was successful
+ * \return ILM_FAILED if the client can not call the method on the service.
+ */
+ilmErrorTypes ilm_takeScreenshotWithCallbacks(t_ilm_uint screen, t_ilm_const_string filename,
```
Comment (efriedrich) [inline ivi-layermanagement-api/ilmControl/include/ilm_control.h:384]: to make the api more clear and avoid parallel access to the memory from Weston and from the client during the accessing the same memory in the done callback I would remove the filename from the api parameters 
then we can rename the api's like ilm_takeAsyncScreenshot

File: ivi-layermanagement-api/ilmControl/include/ilm_control.h [hunk 2]
```diff
@@ -376,6 +397,27 @@ ilmErrorTypes ilm_takeScreenshot(t_ilm_uint screen, t_ilm_const_string filename)
...
+ * \param[in] filename Optional location where the screenshot should be stored,
+ * if the filename is empty and callbacks are available, function will not blocks
+ * \param[in] surfaceid Identifier of the surface to take the screenshot of
+ * \param[in] callback_done Optional callback called when screenshot is acquired
+ * \param[in] callback_error Optional callback called when screenshot acqusition failed
+ * \param[in] user_data Optional callback user data passed in by called
+ * \return ILM_SUCCESS if the method call was successful
+ * \return ILM_FAILED if the client can not call the method on the service.
+ */
+ilmErrorTypes ilm_takeSurfaceScreenshotWithCallbacks(t_ilm_const_string filename, t_ilm_surface surfaceid,
```
Comment (HarshaMM) [inline ivi-layermanagement-api/ilmControl/include/ilm_control.h:416]: I think API becomes a bit more complicated with filename in place. screenshot fd is given to application and application can dump it to a file or process it. Additional encoding to a file is already supported by another API. Application can read and encode to a file. I do not see a need to provide a union of multiple functionalities in one API and make it complex.

My proposal would be remove filename from the argument.

**Comment by khangtb1** [comment]
Thank @efriedrich  and @HarshaMM for your suggestion

I pushed the new code to update your point, the new api only include the things relate to callback, remove the filename for more simple and clear.

File: ivi-layermanagement-api/ilmCommon/include/ilm_types.h [hunk 3]
```diff
@@ -287,4 +287,33 @@ typedef void(*notificationFunc)(ilmObjectType object,
 typedef void(*shutdownNotificationFunc)(t_ilm_shutdown_error_type error_type,
                                         int errornum,
                                         void* user_data);
+
+/**
+ * Typedef for notification callback on screenshot send done event
+ * @param user_data the use data, be passed when call the screenshot api
+ * @param fd fd for file containing image data, don't close it in callback,
+ * it will be closed after the callback execution
```
Comment (efriedrich) [inline ivi-layermanagement-api/ilmCommon/include/ilm_types.h:295]: please add the comment that the FD should not be accessed any longer after the callback execution

Comment (khangtb1) [reply ivi-layermanagement-api/ilmCommon/include/ilm_types.h]: I updated the description. Thanks.
@audoan99 I think this ready to merge. Could you please have check with this?

**Comment by audoan99** [comment]
Tested-by: Au Doan Ngoc [au.doanngoc@vn.bosch.com](mailto:au.doanngoc@vn.bosch.com)

**Review by efriedrich — APPROVED** [review]

**Pull request merged**